### PR TITLE
GAWB-3569 Can create specify Auth Domains on Resource creation

### PR DESCRIPTION
**Pull request by gpolumbo-broad**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/resources/swagger/api-docs.yaml [hunk 1]
```diff
@@ -1122,10 +1122,14 @@ definitions:
         type: string
         description: id of the resource to create
       policies:
-        type: object
-        additionalProperties:
+        type: array
+        items:
           $ref: '#/definitions/AccessPolicyMembership'
         description: map of initial policies to create
+      authDomains:
```
Comment (MatthewBemis) [inline src/main/resources/swagger/api-docs.yaml]: for here and elsewhere in the PR: authDomains should be singular. there's only ever 1 auth domain for a resource, which may be comprised of any number of managed groups

**Comment by gpolumbo-broad** [comment]
jenkins retest

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 2]
```diff
@@ -83,16 +81,45 @@ class ResourceService(private val resourceTypes: Map[ResourceTypeName, ResourceT
...
+      val errors = constrainableErrors ++ existenceErrors.flatten
+      if (errors.nonEmpty) {
+        Option(ErrorReport("Invalid Auth Domains specified", errors))
+      } else None
+    }
+  }
+
+  private def validateAuthDomainsExist(authDomains: Set[WorkbenchGroupName]): Future[Set[Option[ErrorReport]]] = {
+    Future.traverse(authDomains) { groupName =>
+      directoryDAO.loadGroup(groupName) map {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: this should check against managed groups... I think we should add a "use" action to managed groups that both members and admins have. Then this should check that the user has that action. If the user does not have that action then the error message should say that either the group does not exist or the user does not have access.

**Review by dvoet — CHANGES_REQUESTED** [review]
see comments

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala [hunk 3]
```diff
@@ -83,16 +81,44 @@ class ResourceService(private val resourceTypes: Map[ResourceTypeName, ResourceT
...
+    validateAuthDomainPermissions(authDomain, userInfo).map { permissionsErrors =>
+      val constrainableErrors = validateAuthDomainConstraints(resourceType, authDomain).toSeq
+      val errors = constrainableErrors ++ permissionsErrors.flatten
+      if (errors.nonEmpty) {
+        Option(ErrorReport("Invalid Auth Domain specified", errors))
+      } else None
+    }
+  }
+
+  private def validateAuthDomainPermissions(authDomain: Set[WorkbenchGroupName], userInfo: UserInfo): Future[Set[Option[ErrorReport]]] = Future.sequence(
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala]: User Future.traverse instead of sequence and map

**Review by dvoet — APPROVED** [review]

**Review by MatthewBemis — APPROVED** [review]

**Pull request merged**